### ISSUE TITLE
Make widgets as powerful as components

### ISSUE THREAD

**Issue by paf31** [slamdata/purescript-halogen#55]
and update the `ace` example

**Comment by jdegoes** [comment]
In fact, I'd argue that at some level, a widget must be able to be wrapped up into a component, so that it can accept and generate messages exactly like a component. Strictly one-way communication will not work for many use cases and will necessitate messy workarounds.

**Comment by paf31** [comment]
I think it should be possible to wrap widgets as components using some kind of helper function.

The tricky part right now is that I'm running into variance issues trying to make `Widget` into a `Functor`. The type of the renderer becomes:

```purescript
(i -> Eff eff Unit) -> (p -> Widget eff i) -> HTML p i -> VTree
```

which is invariant in `i`. `Widget` essentially has become 

```purescript
type Widget eff i = (i -> Eff eff Unit) -> VTree
```

which is covariant.

Maybe I've just been staring at this for too long, but I feel like these pieces should be fitting together.

**Comment by jdegoes** [comment]
> which is covariant, but this is tricky because of the finally tagless approach.

Well, finally tagless is subservient to good composable abstractions (IMO), so if it proves an obstacle, we can always go back to initial.

**Comment by jdegoes** [comment]
This ticket should really be retitled, "Widgets should be as powerful as Components".

**Comment by paf31** [comment]
Fixed in #58. The Ace editor example is a good demo of this.
